src/main/java/com/nakal/ScreenExecutor/NakalExecutor.java

```diff
@@ -39,107 +37,85 @@ public NakalExecutor(){
      */
     public boolean nakalExecutorNativeCompare(String baseLineImageName)
         throws InterruptedException, IOException, IM4JavaException {
-        return compareTwoImages(baseLineImageName);
+        screenPaths = new ScreenPaths(baseLineImageName);
+        if (isBuildMode()) {
+            return buildMode(baseLineImageName);
+        } else if (isCompareMode()) {
+            return compareMode(baseLineImageName,null);
+        }
+        return false;
     }
-
-    private boolean compareTwoImages(String baseLineImageName)
-        throws InterruptedException, IOException, IM4JavaException {
-        initialize(baseLineImageName);
+    /**
+     * @param baseLineImageName
+     * @param threshold
+     * @return false if actual and expected images are not similar and generate a difference Image
+     */
+    public boolean nakalExecutorNativeCompare(String baseLineImageName, int threshold)
+            throws InterruptedException, IOException, IM4JavaException {
+        screenPaths = new ScreenPaths(baseLineImageName);
         if (isBuildMode()) {
             return buildMode(baseLineImageName);
         } else if (isCompareMode()) {
-            try {
-                screenCaptureAndMaskRegionsIfPresent(baseLineImageName, baseLineImageName,
-                    screenPaths.getActualImage(), screenPaths.getActualMaskedRegionImage(),
-                    screenPaths.getMaskedActualImage());
-
-                if (imageUtil.compareImages(screenPaths.getMaskedExpectedImage(),
-                        screenPaths.getMaskedActualImage(), screenPaths.getDiffImage())) {
-                    return true;
-                } else {
-                    mergerDiffHorizontal();
-                }
-
-            } catch (IOException e) {
-                e.printStackTrace();
-            } catch (InterruptedException e) {
-                e.printStackTrace();
-            } catch (IM4JavaException e) {
-                e.printStackTrace();
-            }
+            return compareMode(baseLineImageName, threshold);
+        }
+        return false;
+    }
+    private boolean compareMode(String baseLineImageName,Integer threshold)
+            throws InterruptedException, IOException, IM4JavaException {
+        String actualImage= screenPaths.getActualImage();
+        String maskedActualImage=screenPaths.getMaskedActualImage();
+        String maskedRegionActualImage= screenPaths.getMaskedRegionActualImage();
+        screenCaptureAndMaskIfExist(baseLineImageName,actualImage,maskedRegionActualImage,maskedActualImage);
+        if (threshold != null) {
```

Comment:
Can integer be `null`?

Reply:
Yes Integer class can be null. There are 2 other ways we can handle this as well. We can use java.util.Optional<Integer> (which is introduced in Java 8)for this case as well or we can pass negative value and handle it accordingly. Pls. suggest

Reply:
Definitely we should use optionals for other null checks at some point in time but i didn't the need of using `Integer` over `int` here. May be we mistakenly did it initially. Ideally we should call compare method with threshold as `0` instead of `null` if user doesn't provide any threshold.